File: app/src/main/java/bjzhou/coolapk/app/ui/adapters/UpgradeAdapter.java
package bjzhou.coolapk.app.ui.adapters;

import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import bjzhou.coolapk.app.R;
import bjzhou.coolapk.app.model.UpgradeApkExtend;
import bjzhou.coolapk.app.net.ApkDownloader;
import bjzhou.coolapk.app.net.DownloadMonitor;

/**
 * Created by bjzhou on 14-8-13.
 */
public class UpgradeAdapter extends RecyclerView.Adapter {

    private static final String TAG = "UpgradeAdapter";
    private final FragmentActivity mActivity;
    private List<UpgradeApkExtend> mUpgradeList = new ArrayList<>();

    public UpgradeAdapter(FragmentActivity activity) {
        mActivity = activity;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View convertView = mActivity.getLayoutInflater().inflate(R.layout.list_item_upgrade_app, viewGroup, false);
        return new ViewHolder(convertView);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
        final ViewHolder holder = (ViewHolder) viewHolder;
        holder.titleView.setText(mUpgradeList.get(position).getTitle());
        holder.infoView.setText(mUpgradeList.get(position).getInfo());
        holder.logoView.setImageDrawable(mUpgradeList.get(position).getLogo());
        String changelog = mUpgradeList.get(position).getApk().getChangelog();
        if (!TextUtils.isEmpty(changelog)) {
            holder.changelogView.setText(changelog);
            holder.changelogView.setVisibility(View.VISIBLE);
        }
        holder.upgradeButton.setTag(0);
        holder.upgradeButton.setText("升级");
        int id = mUpgradeList.get(position).getApk().getId();
        if (ApkDownloader.getInstance().isDownloading(id)) {
            ApkDownloader.getInstance().setListener(id, new DownloadListener(holder.upgradeButton, id));
        }
    }

    private static class DownloadListener implements DownloadMonitor.DownloadListener {

        private static SparseArray<WeakReference<Button>> sButtonMap = new SparseArray<>();
        private final int id;

        public DownloadListener(Button button, int id) {
            button.setTag(id);
            this.id = id;
            sButtonMap.put(id, new WeakReference<>(button));
        }

        private Button getValidButton() {
            WeakReference<Button> buttonRef = sButtonMap.get(id);
            if (buttonRef != null) {
                Button button = buttonRef.get();
                if (button != null && (int) button.getTag() == id) {
                    return button;
                }
            }
            return null;
        }

        @Override
        public void onDownloading(int percent) {
            Button button = getValidButton();
            if (button != null) {
                button.setText(percent + "%");
            }
        }

        @Override
        public void onFailure(int errCode, String... err) {
            Button button = getValidButton();
            if (button != null) {
                button.setText("下载失败,点击重试");
            }
        }

        @Override
        public void onDownloaded() {
            Button button = getValidButton();
            if (button != null) {
                button.setText("正在安装");
            }
        }

        @Override
        public void onComplete() {
            Button button = getValidButton();
            if (button != null) {
                button.setText("安装成功");
            }
        }
    }

    @Override
    public long getItemId(int position) {
        return mUpgradeList.get(position).getApk().getId();
    }

    @Override
    public int getItemCount() {
        return mUpgradeList.size();
    }

    public void setUpgradeList(List<UpgradeApkExtend> upgradeList) {
        mUpgradeList = upgradeList;
    }

    private class ViewHolder extends RecyclerView.ViewHolder {
        ImageView logoView;
        TextView titleView;
        TextView infoView;
        TextView changelogView;
        Button upgradeButton;

        public ViewHolder(View itemView) {
            super(itemView);
            logoView = (ImageView) itemView.findViewById(R.id.list_item_icon);
            titleView = (TextView) itemView.findViewById(R.id.list_item_title);
            infoView = (TextView) itemView.findViewById(R.id.list_item_info);
            changelogView = (TextView) itemView.findViewById(R.id.list_item_description);
            upgradeButton = (Button) itemView.findViewById(R.id.list_item_upgrade);

            upgradeButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    UpgradeApkExtend apkExtend = mUpgradeList.get(getAdapterPosition());
                    if (!ApkDownloader.getInstance().isDownloading(apkExtend.getApk().getId())) {
                        upgradeButton.setText("正在准备下载");
                        apkExtend.getApk().setTitle(apkExtend.getTitle());
                        ApkDownloader.getInstance().downloadAndInstall(mActivity, apkExtend.getApk(),
                                new DownloadListener(upgradeButton, apkExtend.getApk().getId()));
                    }
                }
            });
        }
    }
}

File: app/src/main/kotlin/bjzhou/coolapk/app/ui/adapters/UpgradeAdapter.kt
package bjzhou.coolapk.app.ui.adapters

import android.app.DownloadManager
import android.net.Uri
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import bjzhou.coolapk.app.App
import bjzhou.coolapk.app.R
import bjzhou.coolapk.app.model.Apk
import bjzhou.coolapk.app.model.UpgradeApkExtend
import bjzhou.coolapk.app.net.ApkDownloader
import bjzhou.coolapk.app.net.DownloadStatus
import bjzhou.coolapk.app.util.Settings
import bjzhou.coolapk.app.util.Utils
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.list_item_upgrade_app.view.*
import java.io.File
import java.util.*

/**
 * Created by bjzhou on 14-8-13.
 */
class UpgradeAdapter(private val mActivity: FragmentActivity) : RecyclerView.Adapter<UpgradeAdapter.ViewHolder>() {
    private var mUpgradeList: List<UpgradeApkExtend> = ArrayList()
    private var mObserveMap = HashMap<Apk, Disposable>()
    private val mButtonMap = WeakHashMap<Apk, Button>()

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): UpgradeAdapter.ViewHolder {
        val convertView = mActivity.layoutInflater.inflate(R.layout.list_item_upgrade_app, viewGroup, false)
        return ViewHolder(convertView)
    }

    override fun onBindViewHolder(holder: UpgradeAdapter.ViewHolder, position: Int) {
        val upgradeApk = mUpgradeList[position]
        holder.titleView.text = upgradeApk.title ?: upgradeApk.apk.title ?: "null"
        holder.infoView.text = upgradeApk.info ?: upgradeApk.apk.info ?: "null"
        holder.logoView.setImageDrawable(upgradeApk.logo)
        val changelog = upgradeApk.apk.changelog
        if (!TextUtils.isEmpty(changelog)) {
            holder.changelogView.text = changelog
            holder.changelogView.visibility = View.VISIBLE
        } else {
            holder.changelogView.visibility = View.GONE
        }
        holder.upgradeButton.tag = 0
        holder.upgradeButton.text = "升级"
        holder.upgradeButton.tag = upgradeApk.apk
        mButtonMap.put(upgradeApk.apk, holder.upgradeButton)
    }

    override fun getItemId(position: Int): Long {
        return mUpgradeList[position].apk.id.toLong()
    }

    override fun getItemCount(): Int {
        return mUpgradeList.size
    }

    fun setUpgradeList(upgradeList: List<UpgradeApkExtend>) {
        mUpgradeList = upgradeList
        for (upgradeApk in mUpgradeList) {
            mObserveMap.put(upgradeApk.apk, observeApk(upgradeApk.apk))
        }
    }

    private fun observeApk(apk: Apk): Disposable {
        return ApkDownloader.instance.observeDownloadStatus(apk)
                .subscribe {
                    val button = mButtonMap[apk]
                    if (button != null && button.tag as Apk == apk) {
                        button.post {
                            if (it.status == DownloadStatus.STATUS_NOT_STARTED) {
                                button.text = "升级"
                            } else if (it.status == DownloadManager.STATUS_FAILED) {
                                button.text = "下载失败,点击重试"
                            } else if (it.status == DownloadManager.STATUS_SUCCESSFUL) {
                                button.text = "下载完成"
                            } else {
                                button.text = it.percent.toString() + "%"
                            }
                        }
                    }
                }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal var logoView = itemView.list_item_icon
        internal var titleView = itemView.list_item_title
        internal var infoView = itemView.list_item_info
        internal var changelogView = itemView.list_item_description
        internal var upgradeButton = itemView.list_item_upgrade

        init {
            upgradeButton.setOnClickListener {
                if (adapterPosition == RecyclerView.NO_POSITION) return@setOnClickListener
                val apkExtend = mUpgradeList[adapterPosition]
                val status = ApkDownloader.instance.getDownloadStatus(apkExtend.apk)
                if (status.status == DownloadManager.STATUS_SUCCESSFUL) {
                    val file = File(Settings.instance.downloadDir, apkExtend.apk.filename)
                    val pkgInfo = App.context.packageManager.getPackageArchiveInfo(file.absolutePath, 0)
                    if (file.exists() && pkgInfo != null) {
                        Utils.installApk(Uri.fromFile(file))
                    }
                } else if (status.status == DownloadStatus.STATUS_NOT_STARTED || status.status == DownloadManager.STATUS_FAILED) {
                    ApkDownloader.instance.download(apkExtend.apk)
                    mObserveMap.put(apkExtend.apk, observeApk(apkExtend.apk))
                }
            }
        }
    }

    companion object {

        private val TAG = "UpgradeAdapter"
    }
}
